src/modules/utils/directory_win.rs

```diff
@@ -0,0 +1,113 @@
+extern crate winapi;
+
+use std::ffi::OsStr;
+use std::iter;
+use std::mem;
+use std::os::windows::ffi::OsStrExt;
+use winapi::ctypes::c_void;
+use winapi::shared::minwindef::{BOOL, DWORD};
+use winapi::um::handleapi;
+use winapi::um::processthreadsapi;
+use winapi::um::securitybaseapi;
+use winapi::um::winnt::{
+    SecurityImpersonation, DACL_SECURITY_INFORMATION, FILE_ALL_ACCESS, FILE_GENERIC_EXECUTE,
+    FILE_GENERIC_READ, FILE_GENERIC_WRITE, GENERIC_MAPPING, GROUP_SECURITY_INFORMATION, HANDLE,
+    OWNER_SECURITY_INFORMATION, PRIVILEGE_SET, PSECURITY_DESCRIPTOR, STANDARD_RIGHTS_READ,
+    TOKEN_DUPLICATE, TOKEN_IMPERSONATE, TOKEN_QUERY,
+};
+
+pub fn is_write_allowed(folder_path: &str) -> std::result::Result<bool, &'static str> {
+    let folder_name: Vec<u16> = OsStr::new(folder_path)
+        .encode_wide()
+        .chain(iter::once(0))
+        .collect();
+    let mut length: DWORD = 0;
+
+    let rc = unsafe {
+        securitybaseapi::GetFileSecurityW(
```

Comment:
Alright, we're about to enter the area of the code I know nothing about, so forgive me if I sound like a total rube here. I just want to understand things well enough to where I could (at least try to) modify this if I need to in the future.

First, what is the difference between `GetFileSecurityA` (which I can find documentation for) and `GetFileSecurityW`?

Second, am I correct in saying that this call is only getting information about `length` so that we can allocate a buffer of the correct size?

Finally (and perhaps the gaps in my knowledge of `unsafe` are showing here), is there an implicit coercion that occurs from `&mut` to `*mut`? The crate docs say that you need an `LPDWORD` (i.e. `*mut DWORD`) for the last argument.

Reply:
A suffix = ANSI string, W suffix = Unicode String
https://docs.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings

Reply:
The second question: yes, one can use type coercion to get a raw pointer from a reference in Rust. More info here: https://doc.rust-lang.org/std/primitive.pointer.html